File: Python/EstruturaDeRepeticao/exercicio_20.py
"""
Altere o programa de cálculo do fatorial, permitindo ao usuário calcular o fatorial várias vezes e limitando o fatorial a números inteiros positivos e menores que 16.
"""

while True:
    fatorial = int(input(f"Digite o fatorial (até 16): "))
    if fatorial <= 0 or fatorial > 16:
        print(f"Não é possível fazer o fatorial de {fatorial}.")
        continue

    print(f"{fatorial}!=", end="")
    numero = 1
    for i in range(fatorial, 0, -1):
        numero *= i
        if i == 1:
            print(f"{i}", end="")
        else:
            print(f"{i}.", end="")
    print(f"={numero}")
    repetir = input("Deseja calcular o fatorial de outro número? (s/n): ")
    if repetir.lower() == 's':
        continue
    elif repetir.lower() == 'n':
        break
    else:
        while repetir not in ["s", "n"]:
            print("Digite uma letra válida.")
            repetir = input("Deseja calcular o fatorial de outro número? (s/n): ")
        if repetir.lower() == "n":
            break

File: Java/EstruturaDeRepeticao/Exercicio20.java
// Altere o programa de cálculo do fatorial, permitindo ao usuário calcular o fatorial várias vezes e limitando o fatorial a números inteiros positivos e menores que 16.
package EstruturaDeRepeticao;
import java.util.Scanner;

public class Exercicio20 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        do {
            System.out.print("Digite um número inteiro positivo e menor que 16 para calcular o fatorial (ou um número negativo para sair): ");
            int numero = scanner.nextInt();

            if (numero < 0) {
                break;
            } else if (numero >= 16 || numero < 0) {
                System.out.println("Número inválido! O número deve ser inteiro positivo e menor que 16.");
            } else {
                int fatorial = 1;
                System.out.print(numero + "!=");
                for (int i = numero; i >= 1; i--) {
                    if (i == 1) {
                        System.out.print(i);
                    } else {
                        System.out.print(i+".");
                    }
                    fatorial *= i;
                }

                System.out.print("=" + fatorial + "\n");
            }
        } while (true);

        System.out.println("Programa encerrado.");
        
        scanner.close();
    }
}
